plugins/in_kafka/in_kafka.c

```diff
@@ -50,111 +51,136 @@ static int try_json(mpack_writer_t *writer, rd_kafka_message_t *rkm)
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        if (rkm->err) {
+            ret = flb_log_event_encoder_append_body_cstring(log_encoder, rd_kafka_message_errstr(rkm));
+        }
+        else {
+            ret = flb_log_event_encoder_append_body_null(log_encoder);
+        }
     }
-    else {
-        mpack_write_nil(writer);
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_append_body_cstring(log_encoder, "key");
     }
 
-    mpack_write_cstr(writer, "key");
-    if (rkm->key) {
-        mpack_write_str(writer, rkm->key, rkm->key_len);
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        if (rkm->key) {
+            ret = flb_log_event_encoder_append_body_string(log_encoder, rkm->key, rkm->key_len);
+        }
+        else {
+            ret = flb_log_event_encoder_append_body_null(log_encoder);
+        }
     }
-    else {
-        mpack_write_nil(writer);
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_append_body_cstring(log_encoder, "payload");
     }
 
-    mpack_write_cstr(writer, "payload");
-    if (rkm->payload) {
-        if (try_json(writer, rkm)) {
-            mpack_write_str(writer, rkm->payload, rkm->len);
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        if (rkm->payload) {
+            if (try_json(log_encoder, rkm)) {
+                ret = flb_log_event_encoder_append_body_string(log_encoder, rkm->payload, rkm->len);
+            }
+        }
+        else {
+            ret = flb_log_event_encoder_append_body_null(log_encoder);
         }
     }
-    else {
-        mpack_write_nil(writer);
+
+    if (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        ret = flb_log_event_encoder_commit_record(log_encoder);
+    }
+
+    if (ret != FLB_EVENT_ENCODER_SUCCESS) {
+        flb_log_event_encoder_rollback_record(log_encoder);
     }
 
-    mpack_writer_flush_message(writer);
+    return ret;
 }
 
 static int in_kafka_collect(struct flb_input_instance *ins,
                             struct flb_config *config, void *in_context)
 {
-    mpack_writer_t writer;
-    char *buf;
-    size_t bufsize;
-    size_t written = 0;
+    int ret;
     struct flb_in_kafka_config *ctx = in_context;
+    rd_kafka_message_t *rkm;
 
-    mpack_writer_init_growable(&writer, &buf, &bufsize);
-
-    if (writer.error == mpack_error_memory) {
-        flb_plg_error(ins, "Failed to allocate buffer.");
-        return -1;
-    }
+    ret = FLB_EVENT_ENCODER_SUCCESS;
 
-    while (true) {
-        rd_kafka_message_t *rkm = rd_kafka_consumer_poll(ctx->kafka.rk, 1);
+    while (ret == FLB_EVENT_ENCODER_SUCCESS) {
+        rkm = rd_kafka_consumer_poll(ctx->kafka.rk, 1);
 
         if (!rkm) {
             break;
         }
 
         flb_plg_debug(ins, "kafka message received");
-        process_message(&writer, rkm);
+        ret = process_message(ctx->log_encoder, rkm);
         rd_kafka_message_destroy(rkm);
         rd_kafka_commit(ctx->kafka.rk, NULL, 0);
```

Comment:
Leave a comment here about this line needing to be wrapped in a `ret == FLB_EVENT_ENCODER_SUCCESS` conditional so we can come back to it if it causes trouble.

I don't want you to change it now because I don't really know if committing the kafka record upon failure is a desired behavior or not because that depends on what caused the issue.